plutus-tx/compiler/Language/PlutusTx/Data.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE LambdaCase #-}
+module Language.PlutusTx.Data where
+
+import           Prelude              hiding (fail)
+
+import           Data.Bifunctor       (bimap)
+import           Data.Bitraversable   (bitraverse)
+import           Data.ByteString.Lazy as BSL
+
+import           Control.Monad.Fail
+
+import qualified Codec.CBOR.Term      as CBOR
+import qualified Codec.Serialise      as Serialise
+
+
+-- | A generic "data" type.
+--
+-- The main constructor 'Constr' represents a datatype value in sum-of-products
+-- form: @Constr i args@ represents a use of the @i@th constructor along with its arguments.
+--
+-- The other constructors are various primitives.
+data Data =
+      Constr Integer [Data]
+    | Map [(Data, Data)]
+    | I Integer
+    | B ByteString
+    deriving (Show, Eq, Ord)
```

Comment:
How do we represent an absence of data?  `Constr 0 []`?

Reply:
That would be an empty tuple. There's no equivalent of `Void` at the moment. I think I'm okay with that - `Void` isn't really "data"!

Reply:
I was just wondering what an empty datascript would look like.  I suppose you'd never need to look at it in that case, so maybe it wouldn't matter.  What's the plan with `PendingTx`?  Is that going to have the contents of the outgoing data scripts embedded in it as non-serialised `Data`?  If so, we'll presumably need to put something in there even if there's an empty "script" (I suppose we have to call them something else now), so `()` would seem like the obvious candidate...

Reply:
We've historically represented "no data script" as `()`, so that's all fine. Or another way of putting it: data scripts are mandatory, so if you don't care you can just put `()` there, which is what we do.

Reply:
Obviously not all outputs are script outputs, but this is just a normal datatype as part of `PendingTx`, nothing special.

Reply:
[Later, having thought a bit more.] I suppose it's actually up to the producer of the output what they put in the associated datascript. Every output has to have a datascript, and it has to contain _something_;  Currently we usually use `()` when there's no content, so I guess we'll just continue to do that.  There's no real need for the `Data` type to have an explicit way of saying "no data".